rfqmsg/accept.go

```diff
@@ -0,0 +1,240 @@
+
+	return tlv.MakeStaticRecord(
+		TypeAcceptID, id, recordSize, IdEncoder, IdDecoder,
+	)
+}
+
+func TypeRecordAcceptAskPrice(askPrice *lnwire.MilliSatoshi) tlv.Record {
+	return tlv.MakeStaticRecord(
+		TypeAcceptAskPrice, askPrice, 8, milliSatoshiEncoder,
+		milliSatoshiDecoder,
+	)
+}
+
+func milliSatoshiEncoder(w io.Writer, val interface{}, buf *[8]byte) error {
+	if ms, ok := val.(*lnwire.MilliSatoshi); ok {
+		msUint64 := uint64(*ms)
+		return tlv.EUint64(w, &msUint64, buf)
+	}
+
+	return tlv.NewTypeForEncodingErr(val, "MilliSatoshi")
+}
+
+func milliSatoshiDecoder(r io.Reader, val interface{}, buf *[8]byte,
+	l uint64) error {
+
+	if ms, ok := val.(*lnwire.MilliSatoshi); ok {
+		var msInt uint64
+		err := tlv.DUint64(r, &msInt, buf, l)
+		if err != nil {
+			return err
+		}
+
+		*ms = lnwire.MilliSatoshi(msInt)
+		return nil
+	}
+
+	return tlv.NewTypeForDecodingErr(val, "MilliSatoshi", l, 8)
+}
+
+func TypeRecordAcceptExpiry(expirySeconds *uint64) tlv.Record {
+	return tlv.MakePrimitiveRecord(
+		TypeAcceptExpiry, expirySeconds,
+	)
+}
+
+func TypeRecordAcceptSig(sig *[64]byte) tlv.Record {
+	return tlv.MakePrimitiveRecord(
+		TypeAcceptSignature, sig,
+	)
+}
+
+// acceptMsgData is a struct that represents the data field of a quote
+// accept message.
+type acceptMsgData struct {
+	// ID represents the unique identifier of the quote request message that
+	// this response is associated with.
+	ID ID
+
+	// AskPrice is the asking price of the quote.
+	AskPrice lnwire.MilliSatoshi
+
+	// Expiry is the asking price expiry lifetime unix timestamp.
+	Expiry uint64
+
+	// sig is a signature over the serialized contents of the message.
+	sig [64]byte
+}
+
+// encodeRecords determines the non-nil records to include when encoding at
+// runtime.
+func (q *acceptMsgData) encodeRecords() []tlv.Record {
+	var records []tlv.Record
+
+	// Add id record.
+	records = append(records, TypeRecordAcceptID(&q.ID))
+
+	// Add ask price record.
+	records = append(records, TypeRecordAcceptAskPrice(&q.AskPrice))
+
+	// Add expiry record.
+	records = append(
+		records, TypeRecordAcceptExpiry(&q.Expiry),
+	)
+
+	// Add signature record.
+	records = append(
+		records, TypeRecordAcceptSig(&q.sig),
+	)
+
+	return records
+}
+
+// Encode encodes the structure into a TLV stream.
+func (q *acceptMsgData) Encode(writer io.Writer) error {
+	stream, err := tlv.NewStream(q.encodeRecords()...)
+	if err != nil {
+		return err
+	}
+	return stream.Encode(writer)
+}
+
+// DecodeRecords provides all TLV records for decoding.
+func (q *acceptMsgData) decodeRecords() []tlv.Record {
+	return []tlv.Record{
+		TypeRecordAcceptID(&q.ID),
+		TypeRecordAcceptAskPrice(&q.AskPrice),
+		TypeRecordAcceptExpiry(&q.Expiry),
+		TypeRecordAcceptSig(&q.sig),
+	}
+}
+
+// Decode decodes the structure from a TLV stream.
+func (q *acceptMsgData) Decode(r io.Reader) error {
+	stream, err := tlv.NewStream(q.decodeRecords()...)
+	if err != nil {
+		return err
+	}
+	return stream.Decode(r)
```

Comment:
I think we'd want `stream.DecodeP2P()` for everything in `rfq`.

Reply:
Yes, good point (main difference is not allowing any messages longer than 65k bytes to prevent memory exhaustion attacks).

Reply:
Good to know, thanks.